flask_chat/services.py

```diff
@@ -0,0 +1,270 @@
+"""Services to execute database transactions"""
+
+import os
+import logging
+
+from datetime import datetime
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.exc import SQLAlchemyError
+
+from models import User, Message
+from utils.helpers import random_strings_generator
+from utils.constants import MSG_LOAD_BATCH
+
+# Initialise logger
+logger = logging.getLogger("gunicorn.access")
+
+# Retrieve environment variables
+POSTGRES_USER = os.getenv("POSTGRES_USERNAME")
+POSTGRES_PASSWORD = os.getenv("POSTGRES_PASSWORD")
+POSTGRES_DATABASE = os.getenv("PGDATABASE")
+POSTGRES_HOSTNAME = os.getenv("POSTGRES_HOSTNAME")
+
+# Connect to database
+conn_string = f"postgresql+psycopg2://{POSTGRES_USER}:{POSTGRES_PASSWORD}@{POSTGRES_HOSTNAME}:5432/{POSTGRES_DATABASE}"
+engine = create_engine(conn_string)
+
+
+class UserService:
+    """Operate user-related transactions"""
+
+    def __init__(self):
+        self.session = sessionmaker(engine)
+
+    def get_user_by_id(self, identity):
+        """Return User object by its identifier.
+
+        ## Parameters:
+            **identity** (_str_):
+            User unique identifier.
+
+        ### Returns:
+            _User_:
+            User object. _None_ if exception caught.
+        """
+        try:
+            session = self.session()
+            result = session.query(User).get(identity)
+            session.close()
+            return result
+        except SQLAlchemyError:
+            logger.error("An error occured while establishing conection with PostgreSQL instance.")
+            return None
+
+    def get_user_info(self, **kwargs):
+        r"""Retrieve list of users from connected database. Provide additional
+        arguments to filter results by column (i.e. `WHERE` clause).
+
+        ### Returns:
+            _List\[User\]_:
+            list of User objects. _None_ if an exception is caught.
+        """
+        try:
+            session = self.session()
+
+            if kwargs:
+                results = session.query(User).filter_by(**kwargs).all()
+            else:
+                results = session.query(User).all()
+
+            logger.debug(results)
+            session.close()
+            return results
+        except SQLAlchemyError:
```

Comment:
Please, think about closing session on error. Here and in all such cases

Reply:
Fixed.